### DISCUSSION TITLE
v0.0.1.dev0

### DISCUSSION THREAD

**Discussion by rdbende** [tukaan/tukaan#72] [Announcements]
This is the first release of Tukaan! ✨🥳
Tukaan website: https://tukaan.github.io
Pypi project link: https://pypi.org/project/tukaan

Feedback is welcome!

<hr /><em>This discussion was created from the release <a href='https://github.com/tukaan/tukaan/releases/tag/v0.0.1.dev0'>v0.0.1.dev0</a>.</em>

**Comment by Moosems** [comment]
So excited to see this going to pypi! Can't wait to use it! I am having a problem though where because I'm on a mac that it crashes python when i try running tukaan scripts. Any ideas?

**Reply by lg-nineoneninethreenine** [reply]
This was compiled on Linux, so we're unsure if it works everywhere. However, we are trying to find more ways to get this to almost every platform.

**Reply by Moosems** [reply]
No worries! I just though i'd let you know. I use a mac so if you give detailed instructions on the compilation process I could try doing a mac version for you.  Your thoughts?

**Reply by lg-nineoneninethreenine** [reply]
Hmm... We do have GitHub Actions runtimes, but we need a tester ofc. If @rdbende agrees, you can try checking for compiling Tukaan. My guess is that it would be running the setup file and creating an installable binary. Then, you install it with PIP and test it out. Just make sure you know the module path, since if it malfunctions, you can remove it and test another build.

**Reply by rdbende** [reply]
Will try, how it works on Mac, probably there's a bit more to do, than just build binaries.
I'll try to set up some Mac VM when I can, and try Tukaan on it. (MacOS requires some much storage, that when I last tried to make a VM, my hard drive went crazy 🤣)

> if you give detailed instructions on the compilation process

I can't provide instructions, as  even I don't know them 🙃

**Reply by Moosems** [reply]
> Will try, how it works on Mac, probably there's a bit more to do, than just build binaries. I'll try to set up some Mac VM when I can, and try Tukaan on it. (MacOS requires some much storage, that when I last tried to make a VM, my hard drive went crazy 🤣)

Haha I get that. It's definitely not a lightweight os 😂. It's probably better to buy a completely separate SSD for that if you really want to set up a vm. Honestly it's kinda annoying how much space and processing power it takes to run macos. If theres anything i can do let me know 😉

**Reply by Moosems** [reply]
> Will try, how it works on Mac, probably there's a bit more to do, than just build binaries. I'll try to set up some Mac VM when I can, and try Tukaan on it. (MacOS requires some much storage, that when I last tried to make a VM, my hard drive went crazy 🤣)
> 
> > if you give detailed instructions on the compilation process
> 
> I can't provide instructions, as even I don't know them 🙃

How did you upload it then?

**Reply by rdbende** [reply]
As you may have read above, the build scripts aren't from me, so I don't know if they still work after 18 years. However, I have put together a short tutorial thingy that might work, although I don't have high hopes for it and I make no guarantee.

## Install build dependencies
```bash
$ brew install make
$ brew install tcl-tk-dev
```

## Clone repo and set up submodules
```bash
$ git clone https://github.com/tukaan/tukaan
$ cd tukaan
$ git submodule init
$ git submodule update
```

## Build Serif
```bash
$ cd tukaan/Serif
$ ./build.sh x64
```
Now you should have a file named `serif_mac_x64.dylib` in Serif/pkg

## Build Snack
```bash
$ cd tukaan/Snack/src/unix
$ ./configure --with-tcl=/Library/Frameworks/Tcl.framework --with-tk=/Library/Frameworks/Tk.framework
```
- Now, you have to edit the Makefile and specify the Tcl and Tk include file directory as `/Library/Frameworks/Tcl.framework/Headers` and `/Library/Frameworks/Tk.framework/Headers`

Then:
```bash
$ make
```
~Now you should have a folder inside Snack/pkg named `mac-x64` and a binary file named `snack.dylib` in it.~
EDIT:
Now you should have a binary file named `snack.dylib` in src/unix, please copy it to Snack/pkg/mac-x64.

If you got here, it's a success, I don't expect it to happen, something is definitely broken in the build system 🤪

**Reply by Moosems** [reply]
Is there any way to do this without homebrew? I know its stupid but I've made a point to not install homebrew on my mac and am wondering if there are any alternatives. As for the ```snak.dylib``` is there a way for you to do that in the main system for future builds? Thanks for putting this together by the way! 😁

**Reply by lg-nineoneninethreenine** [reply]
> I'll try to set up some Mac VM when I can, and try Tukaan on it. (MacOS requires some much storage, that when I last tried to make a VM, my hard drive went crazy 🤣)

@rdbende be careful with that. Last time I saw LTT getting it on a rig of his, the image file for Big Sur was 22 GB huge. Colossally big.

**Reply by lg-nineoneninethreenine** [reply]
> Is there any way to do this without homebrew?

Is there any way to get these packages without Homebrew?

**Reply by lg-nineoneninethreenine** [reply]
Wait, macOS kinda has a Unix kernel, so why not test on Unix?

**Reply by Moosems** [reply]
That's what I'm thinking. Since it's based on unix it might be worth a shot

**Reply by lg-nineoneninethreenine** [reply]
Yeah, maybe we could try this: I can search for Docker images that have Unix, then I will try running it on Gitpod, test it out. Hope they have GUI control.

**Reply by rdbende** [reply]
> @rdbende be careful with that. Last time I saw LTT getting it on a rig of his, the image file for Big Sur was 22 GB huge. Colossally big

My Monterey is just GBs 🙂

> Wait, macOS kinda has a Unix kernel, so why not test on Unix?

It definitely needs the desktop environment.

> Is there any way to do this without homebrew?

Probably yes, but Idk what alternative package managers does macOS have.

**Reply by lg-nineoneninethreenine** [reply]
> My Monterey is just GBs 🙂

How many?

> Probably yes, but Idk what alternative package managers does macOS have.

Hmm homebrew is just famous, so alternates are remotely possible.

**Reply by rdbende** [reply]
Lol wtf, how did I miss it? It's actually 16 GB.

**Comment by Moosems** [comment]
Is there an equivalent to the notebook tkinter widget yet? Will you be able to make a terminal widget? Once again, I love your work and am really excited to see this project getting so far

**Reply by rdbende** [reply]
Yes, there's the `tukaan.TabView` widget.

## Usage

```python
import tukaan

# Create window
app = tukaan.App("Test TabView widget")

# Create tabview
tabview = tukaan.TabView(app)
tabview.layout.grid()

# Create tabs
tab_1 = tabview.Tab("Tab 1")
tab_2 = tabview.Tab("Tab 2")

# Add tab contents
tukaan.Button(tab_1, "Button in tab 1").layout.grid()
tukaan.Button(tab_2, "Button in tab 2").layout.grid()

app.run()
```

**Reply by Moosems** [reply]
How do I make the tabs text widgets?

**Reply by rdbende** [reply]
If you mean a static text display, you can use the `Label` widget, just like in Tkinter.

And if you want an editable text box, you can use this:
```python
text_box = tukaan.TextBox(tab_1)
text_box.layout.grid()
```
I'll add some detailed documentation on this widget, as it has really advanced features, and some more fancy stuff.

**Reply by Moosems** [reply]
Oops 😬 . I misread the previous code from the first response and didn't see the ```tukaan.Button(tab_1, "Button in tab 1").layout.grid()
tukaan.Button(tab_2, "Button in tab 2").layout.grid()```.  😅 Thank you for the reposnse though and for your explenation above. I really hope to see this project continue long into the future
